### PR TITLE
JSON Serializable checks for array and structs

### PR DESCRIPTION
**Pull request by ashahab**
Checks for np.ndarray and np.generic for logs.item to fix JSON Serializable issues in tenorflow.keras.RemoteMonitor class #32192

### CODE REVIEW

**Comment by ashahab** [comment]
@googlebot I fixed it.

**Comment by ashahab** [comment]
@googlebot I fixed it.

**Comment by rthadur** [comment]
@ashahab thank you , can you please add some test cases around this ?

**Comment by ashahab** [comment]
@rthadur  thanks for looking at this, yes I'm working on the test-cases.

**Comment by ashahab** [comment]
@googlebot I fixed it.

**Comment by ashahab** [comment]
@rthadur I have added a test.

**Comment by ashahab** [comment]
@rchao

**Comment by ashahab** [comment]
@rchao @rthadur Can I get a review?

File: tensorflow/python/keras/callbacks.py
Comment (rchao) [inline tensorflow/python/keras/callbacks.py]: Can you add a comment why these two types are special cases?

File: tensorflow/python/keras/callbacks_test.py
Comment (rchao) [inline tensorflow/python/keras/callbacks_test.py]: Do those two tests cover both np.ndarray and np.generic cases? Is it possible to assert the `send` dict to make sure the items are expected?

**Review by rchao — COMMENTED** [review]
Thanks for the PR!

File: tensorflow/python/keras/callbacks.py
Comment (ashahab) [reply tensorflow/python/keras/callbacks.py]: Done

File: tensorflow/python/keras/callbacks_test.py
Comment (ashahab) [reply tensorflow/python/keras/callbacks_test.py]: yes, added assertions

**Comment by ashahab** [comment]
@rchao  Did you have other questions on this?

**Review by rchao — DISMISSED** [review]
Looks good to me. Thanks for the contribution!

**Comment by ashahab** [comment]
@rchao Who can merge this?

**Comment by rthadur** [comment]
@ashahab can you please check the build failures ?

**Comment by ashahab** [comment]
@rthadur Checked, there was a python style issue.

**Comment by rthadur** [comment]
Thank you re-ran tests again .

**Comment by ashahab** [comment]
@rthadur not sure why the Android CI buidl area failing.

**Comment by ashahab** [comment]
@rthadur Not sure what else is needed for merging the code.

**Comment by rthadur** [comment]
@ashahab thanks for the changes , @rchao gentle ping to review the changes.

File: tensorflow/python/keras/callbacks_test.py
Comment (rchao) [inline tensorflow/python/keras/callbacks_test.py]: Please fix the typo.

**Review by rchao — DISMISSED** [review]
Looks good.

File: tensorflow/python/keras/callbacks_test.py
Comment (ashahab) [reply tensorflow/python/keras/callbacks_test.py]: Fixed.

**Comment by ashahab** [comment]
@rchao sorry need another review(because of the latest change in comment)

**Review by rchao — DISMISSED** [review]

**Comment by ashahab** [comment]
@rthadur @rchao , is there anything else pending on this?

**Review by rchao — DISMISSED** [review]

**Comment by rthadur** [comment]
@rchao can you please help merging this PR.Thank you

**Comment by ashahab** [comment]
@rchao @rthadur Let me know what needs to be done to merge this.

**Comment by ashahab** [comment]
@rchao @rthadur Can this be merged? Also are there steps in the contributor guideline that I've missed? 
BTW if you guys are in tensorflow world summit, I'd love to meet. I'll be speaking at the summit tomorrow.

**Comment by rchao** [comment]
@rthadur Can you check what else needs to be done for this to be merged? Thanks.

**Comment by rthadur** [comment]
@ashahab sorry about that , you PR has been stuck internally and developers are working on it to get merged.Thanks for your patience.
cc @mihaimaruseac

**Comment by mihaimaruseac** [comment]
There is need for more infrastructure changes. I'll tyr to come back to this early next week, sorry for the delay

File: tensorflow/python/keras/callbacks_test.py
Comment (rthadur) [inline tensorflow/python/keras/callbacks_test.py]: ` File "<embedded module '_launcher'>", line 149, in run_filename_as_main
  File "<embedded module '_launcher'>", line 33, in _run_code_in_main
  File "third_party/tensorflow/python/keras/callbacks_test.py", line 26, in <module>
    import requests
ImportError: No module named requests
-- Forge runner: Test failed with exit code 1 while running`

**Review by rthadur — DISMISSED** [review]

File: tensorflow/python/keras/callbacks_test.py
Comment (ashahab) [reply tensorflow/python/keras/callbacks_test.py]: I added `requests` imports to  `tensorflow/tools/ci_build/install/install_pip_packages.sh` to avoid this. Does that not work? An alternative is to do a try catch around the import and not run the test if requests is not available. However that is a less preferable option as this code would not be tested.

Comment (rthadur) [reply tensorflow/python/keras/callbacks_test.py]: @mihaimaruseac any thoughts on this ?

Comment (mihaimaruseac) [reply tensorflow/python/keras/callbacks_test.py]: I might have to do some extra setup internally. Please remind me about this on Monday if I don't get to it by then

Comment (gbaned) [reply tensorflow/python/keras/callbacks_test.py]: @mihaimaruseac  friendly reminder

**Comment by ashahab** [comment]
Any updates on this?

**Comment by mihaimaruseac** [comment]
Apologies. I didn't get a time to look into this. Will dedicate a day this week to handle this and some other similar issues. Apologies again

**Comment by ashahab** [comment]
Hi Mihai,
Was there any progress with this? Is there a way I can help set this up?
Abin

**Comment by ashahab** [comment]
Happy new year.
Is it possible to merge this PR?

**Comment by mihaimaruseac** [comment]
Hi. sorry for the delay. Got swamped with high priority stuff and forgot to check this.

**Comment by ashahab** [comment]
Hi Mihai,
Anything we can do to help proceed with this task?
Abin

**Comment by mihaimaruseac** [comment]
I think this is good to import now. Thank you and sorry for the delay

**Review by mihaimaruseac — APPROVED** [review]

**Comment by ashahab** [comment]
@mihaimaruseac  many thanks!
@rthadur what's the next step to get this merged?

**Comment by mihaimaruseac** [comment]
This still fails in google with 
```
    import requests
ModuleNotFoundError: No module named 'requests'`
```

I think this needs some more manual work internally :(

**Comment by ashahab** [comment]
@mihaimaruseac What does "manual work internally" mean? Do you want me to change the PR? If I make the requests optional, then this may pass but that'd mean skipping the tests always.

**Comment by mihaimaruseac** [comment]
I'll patch the internally imported change. I'll have an update shortly

**Comment by mihaimaruseac** [comment]
One thing why the tests looked OK here but not internally is because of:

https://github.com/tensorflow/tensorflow/blob/8a4e2041f0e1f1710f9a4c70101906d7c39266a4/tensorflow/python/keras/BUILD#L1211

That tag tells bazel to not run the test if in open source mode (i.e., outside of Google).

**Comment by mihaimaruseac** [comment]
Can you please remove that tag and try running the tests again?

I would assume they would fail because `requests` is not added to `workspace.bzl`. Can you add it there, similar to `six_archive` (for `six`)? Or at least tell me which version of `requests` should we import

**Comment by mihaimaruseac** [comment]
I'm patching this internally. No need to do anything

**Comment by mihaimaruseac** [comment]
Hi.

We cannot take a dependency on `urllib` internally as it depends on `chardet` which is licensed under LGPL. As such, please surround the import with a `try` block instead.

File: tensorflow/python/keras/callbacks_test.py
Comment (mihaimaruseac) [inline tensorflow/python/keras/callbacks_test.py]: This is already imported in a try block on line 60.

File: tensorflow/python/keras/callbacks_test.py
Comment (mihaimaruseac) [inline tensorflow/python/keras/callbacks_test.py]: If possible, make this test pass unconditially if `requests` is `None`.

Otherwise, we might need to drop the tests or rewrite them to not use `requests`.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/callbacks_test.py
Comment (ashahab) [reply tensorflow/python/keras/callbacks_test.py]: Added a guard for this.

File: tensorflow/python/keras/callbacks_test.py
Comment (ashahab) [reply tensorflow/python/keras/callbacks_test.py]: removed

**Comment by ashahab** [comment]
@mihaimaruseac I resolved the issues. Please take a look.

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by ashahab** [comment]
@rthadur @mihaimaruseac  Can we get this merged?

**Comment by mihaimaruseac** [comment]
Sanity test needs fixing

```
FAIL: Found 4 non-whitelisted pylint errors:
tensorflow/python/keras/callbacks_test.py:1354: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).

tensorflow/python/keras/callbacks_test.py:1355: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).

tensorflow/python/keras/callbacks_test.py:1367: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).

tensorflow/python/keras/callbacks_test.py:1368: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
```

**Comment by ashahab** [comment]
That's odd. I don't know why the CI tests on my laptop passed.

**Review by mihaimaruseac — DISMISSED** [review]

File: tensorflow/python/keras/callbacks_test.py
Comment (rthadur) [inline tensorflow/python/keras/callbacks_test.py]: @ashahab here is the internal error , can you please take a look 
`Traceback (most recent call last):
 File "<embedded module '_launcher'>", line 165, in run_filename_as_main
 File "<embedded module '_launcher'>", line 39, in _run_code_in_main
 File "/build/work/google3/runfiles/google3/third_party/tensorflow/python/keras/callbacks_test.py", line 26, in <module>
 import requests
ModuleNotFoundError: No module named 'requests'
-- Forge runner: Test failed with exit code 1 `

**Review by rthadur — DISMISSED** [review]

**Comment by ashahab** [comment]
Apologies. I'll look at this tonight.

**Review by mihaimaruseac — DISMISSED** [review]

**Comment by mihaimaruseac** [comment]
Did you change your address? You might need to sign the CLA again

**Comment by ashahab** [comment]
Nope, no changes in the email address I use to commit, very odd.

File: tensorflow/python/keras/callbacks_test.py
Comment (ashahab) [reply tensorflow/python/keras/callbacks_test.py]: replaced this with a context block after the check

**Comment by ashahab** [comment]
Let me know if there's anything remaining, I don't see any failures in the internal CI builds

File: tensorflow/python/keras/callbacks.py
Comment (caisq) [inline tensorflow/python/keras/callbacks.py]: Do we know for sure that we'll get only scalar-shaped `numpy.ndarray`s? I understand that in many cases, this is just a numpy scalar. But I don't know if it's possible that the value can be a non-scalar shape (e.g., `(1, 2)`) now or in the future. 

cc @karmel @fchollet 

There are two options:
1. You may want to add a check of `len(v.shape) == 0` above. 
2. Call `x.tolist()` instead.

I think 2 is the better option, as it works for scalar and non-scalar shapes alike.

**Review by caisq — CHANGES_REQUESTED** [review]

**Comment by ashahab** [comment]
Can you provide an example?

**Comment by caisq** [comment]
@ashahab I can't provide any concrete example for non-scalar numpy array OTOH. But it's not inconceivable that metrics such as PR curve, confusion matrix will emerge in the future, and those are not scalars. That's why I said it's just advocating for covering possible edge cases. What's the downside of using `v.tolist()`? It's functionally equivalent when the shape is `()`: it returns a python scalar; when it's not shaped as `()`, it doesn't error out with an error like `ValueError: can only convert an array of size 1 to a Python scalar`.

Thank you.

**Comment by ashahab** [comment]
@caisq Thanks for your suggestion! 
If I understand correctly, the example of current working structure is this:
```
 logs = {'loss': 0., "val": np.array(1)}
```
which would get flattened to: 
```
send = {'loss': 0., "epoch": 0, "val": 1}
```
You are suggesting we support a struct like this:
```
 logs = {'loss': 0., "val": np.array([[1, 2], [3, 4]])}
```
This would get flattened to:
```
send = {'loss': 0., "epoch": 0, "val": [[1, 2], [3, 4]]}
```
I have a few questions on this approach:
1. What is the cost of doing a recursive `v.tolist()` call on any shaped ndarray? 
2. Why isn't this supported in Keras: https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/callbacks/callbacks.py#L895

**Comment by caisq** [comment]
@ashahab Ah - I wasn't aware that keras-team / keras uses `.item()`. This is indication that at least on the keras-team/keras side, the contract is that metrics and losses will always be a scalar (at least for now), in which case your change is fine. But tf.keras might be different. But I still don't see any argument against `tolist()`. The cost of `tolist()` compared to `item()` on a scalar ndarray should be negligible, and the function is more general. I'll approve the PR now. Feel free to make the `tolist()` change if you agree with me.

**Review by caisq — DISMISSED** [review]

File: tensorflow/python/keras/callbacks.py
Comment (ashahab) [reply tensorflow/python/keras/callbacks.py]: Followed up in latter conversations.

**Comment by ashahab** [comment]
@caisq  Thank you.
@rthadur Let me know if you want me to rebase or can it be pulled.

**Comment by mihaimaruseac** [comment]
It is always better to rebase before pulling.

**Comment by ashahab** [comment]
@mihaimaruseac  rebased

**Comment by ashahab** [comment]
I have rebased this. I see 3 jobs got stuck in Expected — Waiting for status to be reported state. Can someone help?

**Review by mihaimaruseac — APPROVED** [review]

**Comment by ashahab** [comment]
@mihaimaruseac Seems like tests have passed after your approval. What is the next step for merging this?

**Comment by mihaimaruseac** [comment]
It needs importing inside Google source control, running more tests there and another approval internally.

File: tensorflow/python/keras/callbacks_test.py
Comment (mihaimaruseac) [inline tensorflow/python/keras/callbacks_test.py]: From @qlzh727: Should we expect the value of a here?

Comment (mihaimaruseac) [reply tensorflow/python/keras/callbacks_test.py]: Follow-up: Can we update the value of a to be not 0?

**Comment by mihaimaruseac** [comment]
Should be merged soon

**Comment by ashahab** [comment]
Thanks for the merge. I will work on some of the followup tasks in another PR.

**Pull request merged**
